### ISSUE TITLE
Recalculating taxes of fees when using the hook 'woocommerce_saved_order_items'

### ISSUE THREAD

**Issue by stiletagegmbh** [woocommerce/woocommerce#16938]
In our plugin, we recalculate the taxes of fees and shipping costs of an order so that the rules of german law are applied. (The taxes of fees and shipping costs have to be partially taxed based on the taxes of products in the cart, i.e. if there are products with different tax classes, shipping costs and fees have to be taxed partially with different tax classes - simply spoken)

That works fine on frontend.

But we cannot get a solution in backend when a manual order is created.

We tried to hook into "woocommerce_saved_order_items", e.g. for recalculating the taxes of a fee. 
WooCommerce creates the taxes of a fee with the default tax class.
It seems not to be possible to override this default calculation.

Example:
The id of the default tax class is 1,
the id of a reduced tax class ist 3.

To simplyfiy, we want to add a fee, with a tax that consists of
0.5 EUR of tax class 1
as well as
0.25 EUR of tax class 3,

(in practice, these sums are calculated according to the rules of german law)

then we try the following code, hooked into 'woocommerce_saved_order_items'

```
/**
 * Calculating the split tax for fee on ajax callback in backend on "update tax"/"update sum"
 *
 * @wp-hook	woocommerce_saved_order_items
 *
 * @param	int $order_id
 * @return	void
 */
public static function re_calculate_tax_on_save_order_items( $order_id ) {

	$order = wc_get_order( $order_id );

	// getting all fees and remove them from order
	$all_fees = $order->get_fees();
	$order->remove_order_items( 'fee' );

	// loop through all shipping fees and create new ones with the split tax
	foreach ( $all_fees as $fee ) {

		// WC 3.0
		$new_fee            = new WC_Order_Item_Fee();

		$new_taxes = array();
		$new_taxes[ 'total' ] = array();

		$new_taxes[ 'total' ][ 1 ] = 0.5;   // simplyfied for example
		$new_taxes[ 'total' ][ 3 ] = 0.25;

		$new_fee->set_name( $fee->get_name() );
		$new_fee->set_tax_status( 'taxable' );
		$new_fee->set_tax_class( '' );
		$new_fee->set_total( $fee->get_total() );
		$new_fee->set_total_tax( $taxes[ 'sum' ] );
		$new_fee->set_taxes( $new_taxes );

		$new_fee->apply_changes();

		// adding the new fee to order
		$order->add_item( $new_fee );
		$order->save();
	}

}
```

When clicking on the "Recalculate" button with this code,
the taxes for tax class 3 are added correctly, but the taxes for tax class 1 is still the default tax class added by WooCommerce.

We cannot find a solution for our problem.

Maybe you can tell us, if this is a bug or if we are doing something wrong or if we should use another hook to recalculate the taxes?

**Comment by mikejolley** [comment]
If custom code cannot be applied, thats does not make this a bug :) Just a limitation. Taxes are recalculated by core, so to avoid that you'd probably need a new action or filter hook.

Orders call tax calculation functions here: https://github.com/woocommerce/woocommerce/blob/45debc0b4741514f1cffa32710cfecdd29a08057/includes/abstracts/abstract-wc-order.php#L1284-L1302

This in turn fires off calculations per item here for example: https://github.com/woocommerce/woocommerce/blob/56e120790319c885fad6d07ddc3f3a38c4e42af9/includes/class-wc-order-item-fee.php#L71

Maybe a filter could be added there on the taxes before they get set?

Feel free to adapt the core code and send a PR once you have everything working. We can then review it for inclusion. It makes sense for you to do this because you can test the new code before we make any changes.

Thanks

**Comment by stiletagegmbh** [comment]
Thats's my PR:
https://github.com/woocommerce/woocommerce/pull/16969

If it will be approved, will it be included in 3.2?
